### ISSUE TITLE
Does FastAI model usually not train well on regression data?

### ISSUE THREAD

**Issue by SangbumChoi** [awslabs/autogluon#1158]
According to the result of benchmark in kaggle dataset (mercedes, allstate, houses, etc...)

I think that fastainn model does not well-training (which means early stopping in 0 epoch with 20 patience) in regression kaggle datasets (except for allstate).

Since mxnet models usually train well on various regression and binary datasets and also fastainn models train well on binary and multi-class dataset. I assumed that loss normalization or scaling on regression dataset was the main reason of this phenomena (or learning-rate problem). Because as far as I searched fastainn library there was no such functions (ex: loss_scaling_factor) but mxnet does. However, all of the custom trial to enhance fastainn model in regression dataset failed.

Does my experiment (with default hyperparameter such as time, auto-stack, etc...) went wrong or does your group have faced this same problem?

**Comment by SangbumChoi** [comment]
0
TensorBase(5.7082e+08)
1
TensorBase(1.3430e+10)
2
TensorBase(2.8899e+11)
3
TensorBase(3.9408e+12)
4
TensorBase(3.4854e+13)
5
TensorBase(2.3243e+14)
6
TensorBase(6.1357e+14)
7
TensorBase(8.1096e+14)
8
TensorBase(4.5594e+13)
9
TensorBase(5.2544e+13)
10
TensorBase(1.4245e+13)
11
TensorBase(1.0146e+14)
12
TensorBase(2.2767e+14)
13
TensorBase(8.1047e+12)
14
TensorBase(1.3270e+14)
15
TensorBase(3.3033e+14)
16
TensorBase(1.0559e+14)
17
TensorBase(9.1204e+13)
18
TensorBase(1.6465e+14)
19
TensorBase(3.5654e+14)
20
TensorBase(3.4201e+14)

This is the example of Fastainn loss with respective epoch in mercedes dataset

**Comment by Innixma** [comment]
Thanks for noticing this! This appears to be a serious bug that we weren't aware of. At some point it must have been introduced as I remember a long time ago FastAI working correctly in regression problems.

Code showing the bug:

```
from autogluon.tabular import TabularDataset, TabularPredictor
from autogluon.core.utils import generate_train_test_split

LABEL = 'y'
SAMPLE = 1000


if __name__ == '__main__':
    path_prefix = 'datasets/mercedes-benz-greener-manufacturing/'
    path_train = path_prefix + 'train.csv'
    path_test = path_prefix + 'test.csv'

    data = TabularDataset(path_train)
    leaderboard_data = TabularDataset(path_test)

    # train_data = data
    train_data, test_data, train_labels, test_labels = generate_train_test_split(X=data.drop(LABEL, axis=1), y=data[LABEL], problem_type='regression', test_size=0.2)
    train_data[LABEL] = train_labels
    test_data[LABEL] = test_labels

    if SAMPLE is not None and SAMPLE < len(train_data):
        train_data = train_data.sample(n=SAMPLE, random_state=0).reset_index(drop=True)

    ignored_columns = ['ID']

    predictor = TabularPredictor(
        label=LABEL,
        learner_kwargs={'ignored_columns': ignored_columns},
        verbosity=3,
    ).fit(
        train_data=train_data,
        hyperparameters={
            'GBM': {},
            'NN': {},
            'FASTAI': {},
        }
    )

    leaderboard = predictor.leaderboard(test_data)
```

```
Better model found at epoch 0 with _rmse value: 9325.2880859375.
No improvement since epoch 0: early stopping
Model validation metrics: 9325.2880859375
```

Leaderboard:

```
                 model   score_test    score_val  pred_time_test  pred_time_val   fit_time  pred_time_test_marginal  pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  fit_order
0             LightGBM   -10.094900    -7.909521        0.031060       0.024788   0.533952                 0.031060                0.024788           0.533952            1       True          1
1  WeightedEnsemble_L2   -10.096217    -7.870122        1.547460       0.234819  13.676668                 0.024624                0.000724           0.798873            2       True          4
2       NeuralNetMXNet   -10.889210    -8.499052        1.491776       0.209307  12.343843                 1.491776                0.209307          12.343843            1       True          3
3      NeuralNetFastAI -3669.019431 -9325.287763        1.363877       0.120494   4.059517                 1.363877                0.120494           4.059517            1       True          2
```

**Comment by SangbumChoi** [comment]
The two things that I've mentioned will be useful to debug this error (I'm also searching the reason + I expect this is not error because of AutoGluon but FastAI).
1. allstate regression dataset works fine
2. the loss is too large compare to MXNet

**Comment by Innixma** [comment]
#1162 should fix the issue. Running on that branch, I get the following result:

```
                 model  score_test  score_val  pred_time_test  pred_time_val   fit_time  pred_time_test_marginal  pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  fit_order
0             LightGBM  -10.094900  -7.909521        0.029116       0.030299   0.484345                 0.029116                0.030299           0.484345            1       True          1
1  WeightedEnsemble_L2  -10.124198  -7.817343        0.928249       0.279622  15.115147                 0.003840                0.000689           0.196103            2       True          4
2      NeuralNetFastAI  -11.033612  -8.626837        0.308750       0.106975   4.425922                 0.308750                0.106975           4.425922            1       True          2
3       NeuralNetMXNet  -11.039835  -8.364145        0.586542       0.141659  10.008776                 0.586542                0.141659          10.008776            1       True          3
```

There were multiple bugs:

1. We were not applying scaling to the target variable
2. After splitting the data into train and val, some features in train had only 1 unique value. When a row in val has a different unique value, that row will be predicted as a ridiculous value, which is what caused the extremely large error.

These have both been fixed. Thanks again for highlighting the bug!

I will close the issue once the PR is merged.

**Comment by SangbumChoi** [comment]
> #1162 should fix the issue. Running on that branch, I get the following result:
> 
> ```
>                  model  score_test  score_val  pred_time_test  pred_time_val   fit_time  pred_time_test_marginal  pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  fit_order
> 0             LightGBM  -10.094900  -7.909521        0.029116       0.030299   0.484345                 0.029116                0.030299           0.484345            1       True          1
> 1  WeightedEnsemble_L2  -10.124198  -7.817343        0.928249       0.279622  15.115147                 0.003840                0.000689           0.196103            2       True          4
> 2      NeuralNetFastAI  -11.033612  -8.626837        0.308750       0.106975   4.425922                 0.308750                0.106975           4.425922            1       True          2
> 3       NeuralNetMXNet  -11.039835  -8.364145        0.586542       0.141659  10.008776                 0.586542                0.141659          10.008776            1       True          3
> ```
> 
> There were multiple bugs:
> 
> 1. We were not applying scaling to the target variable
> 2. After splitting the data into train and val, some features in train had only 1 unique value. When a row in val has a different unique value, that row will be predicted as a ridiculous value, which is what caused the extremely large error.
> 
> These have both been fixed. Thanks again for highlighting the bug!
> 
> I will close the issue once the PR is merged.

Yeah, I agree with the reason 1.
I think 2 is happening beacuse of subsampling with small portion of data.

However, it is interesting that total score_test got even worst. Maybe different for whole dataset with various models.

Thanks

**Comment by Innixma** [comment]
The worse score_test is likely just random chance and using a small sample of data.

The fix has been merged and will be available in tomorrow's pre-release or by building from source.

Thanks again for spotting this!
